internal/client/consensus/grandpa/communication_gossip.go

```diff
@@ -0,0 +1,1708 @@
+	var catchUpConfig catchUpConfig[N]
+	if config.ObserverEnabled {
+		panic("we have not implemented the observer")
+	} else {
+		// if the observer protocol isn't enabled and we're not a light client, then any full node should be able to
+		// answer catch-up requests.
+		catchUpConfig = catchUpConfigEnabled[N]{false}
+	}
+
+	return inner[H, N, Hasher]{
+		localView:       nil,
+		peers:           newPeers[N](neighborRebroadcastPeriod),
+		liveTopics:      newKeepTopics[H, Hasher](),
+		nextRebroadcast: time.Now().Add(rebroadcastAfter),
+		authorities:     make([]primitives.AuthorityID, 0),
+		pendingCatchUp:  pendingCatchUpNone{},
+		catchUpConfig:   catchUpConfig,
+		config:          config,
+	}
+}
+
+// Note a round in the current set has started. Does nothing if the last call to the function was with the same round.
+func (i *inner[H, N, Hasher]) noteRound(round Round) *peerIDsNeighborPacket[N] {
+	if i.localView.round == round {
+		// Do not send neighbor packets out if `round` has not changed, such behavior is punishable.
+		return nil
+	}
+
+	setID := i.localView.setID
+
+	logger.Debugf("Voter %s noting beginning of round (%d, %d) to network.", i.config.name(), round, setID)
+
+	i.localView.updateRound(round)
+
+	i.liveTopics.push(round, setID)
+	i.peers.reshuffle()
+
+	return i.multicastNeighborPacket()
+}
+
+// Note that a voter set with given ID has started. Does nothing if the last call to the function was with the same
+// SetID.
+func (i *inner[H, N, Hasher]) noteSet(setID SetID, authorities []primitives.AuthorityID) *peerIDsNeighborPacket[N] {
+	if i.localView == nil {
+		i.localView = newLocalView[N](setID, 1)
+	} else {
+		if i.localView.setID == setID {
+			a := make(map[primitives.AuthorityID]struct{})
+			b := make(map[primitives.AuthorityID]struct{})
+
+			for _, auth := range i.authorities {
+				a[auth] = struct{}{}
+			}
+			for _, auth := range authorities {
+				b[auth] = struct{}{}
+			}
+
+			diffAuthorities := !maps.Equal(a, b)
+
+			if diffAuthorities {
+				logger.Debugf(
+					"Gossip validator noted set %s twice with differnet authorities. "+
+						"Was the authority set hard forked?", setID)
+
+				i.authorities = authorities
+			}
+
+			// Do not send neighbor packets out if the `setID` has not changed. Such behavior is punishable.
+			return nil
+		}
+	}
+
+	i.localView.updateSet(setID)
+	i.liveTopics.push(1, setID)
+	i.authorities = authorities
+
+	return i.multicastNeighborPacket()
+}
+
+// Note that we've imported a commit finalizing a given block. Does nothing if the last call to the function was with
+// the same or higher `finalized` number. `setID` & `round` are the ones the commit message is from.
+func (i *inner[H, N, Hasher]) noteCommitFinalized(round Round, setID SetID, finalized N) *peerIDsNeighborPacket[N] {
+	if i.localView == nil {
+		return nil
+	}
+	lch := i.localView.lastCommitHeight()
+	if lch == nil || lch != nil && (*lch < finalized) {
```

Comment:
if lch is nil than the next condition is not evaluated, if lch is not nil then we can evaluate the sec cond safely
```suggestion
	if lch == nil || (*lch < finalized) {
```